### PR TITLE
fixes sharing to group ids with characters that are being url encoded

### PR DESCRIPTION
**Pull request by blizzz**
When you try to share a calendar to (local) group "Route#44" it will fail, because internally the group name was not decoded.

What also was and still is not the case that on failed sharing a proper error was being reported. I.e. at the moment you do not see that it was failing,  only with a reload (or  a complaint by one of the members :)). That might be a follow up PR.

### CODE REVIEW

File: apps/dav/lib/DAV/Sharing/Backend.php [hunk 1]
```diff
@@ -107,6 +107,7 @@ private function shareWith($shareable, $element) {
 			return;
 		}
 
+		$principal[2] = urldecode($principal[2]);
```
Comment (nickvergessen) [inline apps/dav/lib/DAV/Sharing/Backend.php:110]: This breaks groups with a plus as it replaces them with ` ` even if it might not be needed?

Comment (blizzz) [reply apps/dav/lib/DAV/Sharing/Backend.php:110]: Did you try this? Groups containing plusses should suffer the exact same issue. 

But I think it reveals a different problem here … with and without this change groups containing plusses are not being offerend in the calendar share dialogue.

Comment (blizzz) [reply apps/dav/lib/DAV/Sharing/Backend.php:110]: Ok, i see the issue now.

Comment (blizzz) [reply apps/dav/lib/DAV/Sharing/Backend.php:110]: @nickvergessen I added another commit fixing it. But as said, also without the first commit and the calendar changes, a search for groups containing a plus would end up with

```xml
    <?xml version="1.0" encoding="utf-8"?>
    <d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
      <s:exception>Sabre\DAV\Exception\NotFound</s:exception>
      <s:message>Principal with name POW+WOW not found</s:message>
    </d:error>
```

**Comment by blizzz** [comment]
I was finally having another look into this and had to do two more changes (also for carddav) essentially reverting https://github.com/nextcloud/server/pull/5298 which ironically was a fix for the same problem. I suppose it broke somewhere else.

For me it works now with both contacts as well as addressbooks, and I tested with several groups:

Screenshot_20201216_224709

I shared addressbook as well as calendars to these groups.

Screenshot_20201216_224804

Screenshot_20201216_224832

(here you also see that contacts has presentation issues - needs a fix like calendar got)

and all were received:

Screenshot_20201216_224928

Screenshot_20201216_225009

In the database URIs are saved with encoded name (as they should):

Screenshot_20201216_225038

My changes do not! change the behaviour of how the URIs are stored. What changes is the group search (Sharing/Backend changes) and the return of URIs in an properly encoded manner (other changes).  Therefore I do not expect any effects on old shares, even though I do not have deep knowledge of the changes in between in these components or these components in itself.

**Review by nickvergessen — APPROVED** [review]
🐘

**Review by ChristophWurst — APPROVED** [review]
Code looks good

**Pull request merged**